src/libraries/System.Reflection.Emit.ILGeneration/ref/System.Reflection.Emit.ILGeneration.cs

```diff
@@ -13,48 +13,56 @@ public partial class CustomAttributeBuilder
         public CustomAttributeBuilder(System.Reflection.ConstructorInfo con, object?[] constructorArgs, System.Reflection.PropertyInfo[] namedProperties, object?[] propertyValues) { }
         public CustomAttributeBuilder(System.Reflection.ConstructorInfo con, object?[] constructorArgs, System.Reflection.PropertyInfo[] namedProperties, object?[] propertyValues, System.Reflection.FieldInfo[] namedFields, object?[] fieldValues) { }
     }
-    public partial class ILGenerator
+    public abstract class ILGenerator
     {
-        internal ILGenerator() { }
-        public virtual int ILOffset { get { throw null; } }
-        public virtual void BeginCatchBlock(System.Type? exceptionType) { }
-        public virtual void BeginExceptFilterBlock() { }
-        public virtual System.Reflection.Emit.Label BeginExceptionBlock() { throw null; }
-        public virtual void BeginFaultBlock() { }
-        public virtual void BeginFinallyBlock() { }
-        public virtual void BeginScope() { }
+        protected ILGenerator() { }
+        protected abstract void EnsureCapacity(int size);
+        protected abstract void ILEmit(System.Reflection.Emit.OpCode opcode);
+        protected abstract void WriteByte(byte arg);
+        protected abstract void WriteShort(short arg);
+        protected abstract void WriteInt(int arg);
+        protected abstract void WriteLong(long arg);
+        protected abstract void WriteSingle(float arg);
+        protected abstract void WriteDouble(double arg);
```

Comment:
These public APIs were not approved. I do not think we should be adding them. Instead, make more of the methods abstract.

Reply:
> The protected constructor and changing the type to abstract suggests that some of the virtual members should themselves flip to abstract (`or that there's a protected abstract missing from the design`). As a general reminder, abstract methods can't be added after the initial release where it becomes possible to actually extend the type, so for ILGenerator that means the current (.NET 8) release.

From the above API review comment I thought `or that there's a protected abstract missing from the design` part imply that we can add `protected abstract` members, anyway, I have no strong opinion to keep them